lib/Core/BuildEngine.cpp

```diff
@@ -1292,37 +1294,23 @@ class BuildEngineImpl {
     return currentTimestamp;
   }
 
-  KeyID getKeyID(const KeyType& key) {
-      std::lock_guard<std::mutex> guard(keyTableMutex);
-
-    // Delegate if we have a database.
-    if (db) {
-      // Check our cache, to avoid query the database redundantly.
-      auto it = keyTable.find(key);
-      if (it != keyTable.end())
-        return it->second;
-      
-      std::string error;
-      KeyID id = db->getKeyID(key, &error);
-      if (!error.empty()) {
-        delegate.error(error);
-        cancelRemainingTasks();
-      }
-
-      // Cache the result, and return;
-      keyTable.insert(std::make_pair(key, id));
-      return id;
-    }
+  virtual KeyID getKeyID(const KeyType& key) override {
+    std::lock_guard<std::mutex> guard(keyTableMutex);
 
-    // Otherwise, use our builtin key table.
-    //
     // The RHS of the mapping is actually ignored, we use the StringMap's ptr
     // identity because it allows us to efficiently map back to the key string
     // in `getRuleInfoForKey`.
     auto it = keyTable.insert(std::make_pair(key, 0)).first;
     return (KeyID)(uintptr_t)it->getKey().data();
   }
 
+  virtual KeyType getKeyForID(KeyID key) override {
```

Comment:
Should we make this return a const reference instead? That should be a little more efficient, since we are probably copying these keys a bunch right now.

Reply:
Actually, ignore that suggestion... that doesn't work. We would need to return a `StringRef` and that breaks the abstraction of "KeyType".

It's probably worth us having a follow on bug to improve this at some later date though.